src/material/form-field/form-field.ts

```diff
@@ -473,10 +478,10 @@ export class MatFormField
    * checking every change detection cycle.
    */
   private _initializeOutlineLabelOffsetSubscriptions() {
-    // Whenever the prefix changes, schedule an update of the label offset.
-    this._prefixChildren.changes.subscribe(
-      () => (this._needsOutlineLabelOffsetUpdateOnStable = true),
-    );
+    // Whenever the prefix changes, schedule an update of the prefix resize observer for label offset.
+    this._prefixChildren.changes.subscribe(() => {
+      this._needsOutlineLabelPrefixResizeObserverUpdateOnStable = true;
```

Comment:
Rather than observing whenever the child count changes, why don't we do something like this?

```ts
this._hasPrefixChildren = this._prefixChildren.changes
    .pipe(map(prefix => !!prefix.length), distinctUntilChanged());

...

this._hasPrefixChildren.pipe(filter(hasPrefix => hasPrefix)).subscribe(() => {
  this._ngZone.onStable.pipe(take(1), takeUntil(this._destroyed))
      .subscribe(() => this._updateOutlineLabelOffsetOnPrefixResize())
});
```

This way we're only observing when the prefix container pops into existence (i.e. when the number of children goes from 0 to >0).

On the `_resizeObserver` subscription we can `takeUntil(this._hasPrefixChildren)` to stop observing if the children drops back to 0

Reply:
Thank you for mentioning this. the code had an issue about this case. please take another look!

When number of prefix containers goes from n to 0, we need to recalculate the label offset. Otherwise, the label will be misaligned because of the previous `translateX` value.